tower-http/src/limit.rs

```diff
@@ -0,0 +1,246 @@
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<B> Clone for LengthLimitedLayer<B> {
+    fn clone(&self) -> Self {
+        Self {
+            limit: self.limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<B> Copy for LengthLimitedLayer<B> {}
+
+impl<B> fmt::Debug for LengthLimitedLayer<B> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LengthLimitedLayer")
+            .field("body", &any::type_name::<B>())
+            .field("limit", &self.limit)
+            .finish()
+    }
+}
+
+impl<B, S> Layer<S> for LengthLimitedLayer<B> {
+    type Service = LengthLimited<S, B>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        LengthLimited {
+            inner,
+            limit: self.limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+/// Middleware that intercepts requests with body lengths greater than the
+/// configured limit and converts them into `413 Payload Too Large` responses.
+///
+/// See the [module docs](self) for an example.
+pub struct LengthLimited<S, B> {
+    inner: S,
+    limit: usize,
+    _ty: PhantomData<fn() -> B>,
+}
+
+impl<S, B> LengthLimited<S, B> {
+    define_inner_service_accessors!();
+
+    /// Create a new `LengthLimited` with the given body length limit.
+    pub fn new(inner: S, limit: usize) -> Self {
+        Self {
+            inner,
+            limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<S, B> Clone for LengthLimited<S, B>
+where
+    S: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            limit: self.limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<S, B> Copy for LengthLimited<S, B> where S: Copy {}
+
+impl<S, B> fmt::Debug for LengthLimited<S, B>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LengthLimited")
+            .field("inner", &self.inner)
+            .field("service", &format_args!("{}", any::type_name::<B>()))
+            .finish()
+    }
+}
+
+impl<ReqBody, ResBody, S> Service<Request<ReqBody>> for LengthLimited<S, ReqBody>
+where
+    S: Service<Request<Limited<ReqBody>>, Response = Response<ResBody>>,
+    S::Error: Into<BoxError>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+    ResBody::Error: Into<BoxError>,
+{
+    type Response = Response<UnsyncBoxBody<Bytes, BoxError>>;
+    type Error = BoxError;
+    type Future = ResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(|e| e.into())
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let (parts, body) = req.into_parts();
+        let body = Limited::new(body, self.limit);
+        let req = Request::from_parts(parts, body);
+
+        ResponseFuture {
+            future: self.inner.call(req),
+        }
+    }
```

Comment:
I think we should also check if there is `content-length` header and return `413 Payload Too Large` if its larger than the limit.

Reply:
That should be relatively simple to do. It will be a bit simpler than implemented in `axum` because we won't need to worry about disallowing requests lacking a `Content-Length` header.

Reply:
Are you saying you'll look into it?

Reply:
Yes. I'll incorporate this into the layer.

Reply:
This has been done. I've also added some doctests that demonstrate the various cases.

Reply:
I added a fourth usage example for those that do not want to have automatic 413 responses, pointing to use `MapRequestBody` to wrap the inner body with `Limited`.